### DISCUSSION TITLE
Configure OAuth2 access token for authorization in Windows Desktop Application

### DISCUSSION THREAD

**Discussion by ratataplan** [fattureincloud/api#15] [Q&A] [answered]
Is it possible to Configure OAuth2 access token for authorization in Windows Desktop Application?
How can I do?

**Comment by valmoz** [comment, marked as answer]
Hi Carlo,
unfortunately, in our team we don't have expertise with Windows Desktop applications.

Generally speaking, if you want to use the [OAuth2.0 Authorization Code Flow](https://developers.fattureincloud.it/docs/authentication/code-flow/) you must be able to do three things:
- perform REST calls
- open a web page in the browser, to redirect the user to the authorization page
- have a backend to manage the secrets and the redirect URL

If you're able to perform these three operations, then you can use the OAuth2.0 Authorization Code Flow from your environment.

On our documentation, you can find code examples for both our [SDKs](https://developers.fattureincloud.it/docs/authentication/code-flow/sdk-oauth-helper) or with [general purpose code](https://developers.fattureincloud.it/docs/authentication/code-flow/vanilla-code). Feel free to adapt them to your needs.

**In my opinion, this authentication method is not ideal for your use case.** I would suggest checking the other [Authentication methods](https://developers.fattureincloud.it/docs/basics/authentication) to find the one that best suits your needs.

Let us know if you have further questions.

Thanks
Mauro

**Reply by ratataplan** [reply]
Sono riuscito utilizzando OAuth2.0 Authorization Code Flow. Ogni volta che scadono le 24 ore l'utente deve però fare nuovamente l'autenticazione con perdita di tempo ma tutto il processo rimane all'interno dell'applicazione Desktop e l'utente deve solo dare l'autorizzazione in una finestra del browser e dopo aver salvato il token può utilizzare le API 2.0 dalla sua applicazione Desktop Windows.

**Reply by valmoz** [reply]
Buongiorno @ratataplan 
le rispondo in italiano per questa volta, ma in generale cerchiamo di utilizzare l'inglese su questa community.

Ho alcune domande sulla sua implementazione del flusso OAuth:
- Dove ha salvato il client secret? all'interno del codice dell'applicazione desktop?
- Perché non sta utilizzando il refresh token alla scadenza del token principale?
- Ha preso in considerazione gli altri due metodi di autenticazione messi a disposizione, che potrebbero essere più adatti al suo caso d'uso di applicazione desktop?

Resto in attesa di una sua risposta.

Grazie
Mauro

**Reply by ratataplan** [reply]
Yes thank you,
I implemented the method RefreshToken and it works fine without delay time for the user. I did not yet try the others authentication method but I will do soon. The client secret at the moment is inside the code for testing but I will put it in a db.

Thanks
Carlo

**Reply by valmoz** [reply]
Hi @ratataplan,
can you please describe to us how you plan to implement the whole OAuth flow?

Thanks
Mauro

**Reply by valmoz** [reply]
Hi @ratataplan,
I asked to explain your OAuth flow because if not implemented correctly it could bring to potential security issues.
In particular, an attacker could steal your Client ID - Client Secret couple and impersonate your application to perform OAuth requests and authenticate as the users to our APIs. 
That's why it is fundamental that the OAuth flow is implemented correctly and used only when the client secret can be managed properly.

Please contact us back. If you don't want to share your implementation process on a public discussion, you can also contact our Customer Support.

Thanks
Mauro

**Reply by ratataplan** [reply]
Here is how I implemented my class OAuth2
https://github.com/ratataplan/FattureinCloud-OAuth2-flow

**Comment by valmoz** [comment]
Hi @ratataplan
unfortunately, it is not ok.

**The Client Secret must NEVER transit into the customer's PC**, not even if recovered from a DB.
It must be used only from a backend contacted by your windows desktop application (without passing it to the desktop app), or not used at all.

If you don't have a remote backend you can use, you must adopt the [OAuth Device Code Flow](https://developers.fattureincloud.it/docs/authentication/device-code) or [Manual Authentication](https://developers.fattureincloud.it/docs/authentication/manual-authentication).

The three authentication methods use the retrieved token in the same way, so it is really easy to switch to one of the other Auth methods. The difference is that you won't have to manage redirects and secrets.

You should adopt the new Authorization Flow **as soon as possible**. We can support you in the transition if you need help.

Thanks
Mauro

**Comment by ratataplan** [comment]
As backend I could use a sever side application, but in any case the client application need the accessToken and the companyID.

**Reply by valmoz** [reply]
Before reading my comments above: **_Please, note that I still think that the OAuth Device Code is much more easily implementable for you than my proposal below, and (if possible) I would like to know why are you rejecting it_**. It is a flow designed to be used in this particular situation, so it's the best suited to be adopted by a desktop app.

If you still think your best solution is the Authorization Code, you can implement the following flow:
- First, the user clicks on the "connect to Fatture in Cloud" button on your desktop app
- The Desktop app [redirects the user to the FIC Authorization Page opening a browser](https://developers.fattureincloud.it/docs/authentication/code-flow/vanilla-code#1%EF%B8%8F%E2%83%A3-redirect-the-user-to-the-fatture-in-cloud-authorization-page)
- As redirect URL param of the app, you can set up the endpoint to a dedicated page exposed by your Server Side App
- When the user completes the auth step on the FIC page, he will be redirected to the server-side app by our page
- Your Server Side Application [retrieves the get params](https://developers.fattureincloud.it/docs/authentication/code-flow/vanilla-code#2%EF%B8%8F%E2%83%A3-obtain-the-authorization-code) and performs the [Access Token Request](https://developers.fattureincloud.it/docs/authentication/code-flow/vanilla-code#3%EF%B8%8F%E2%83%A3-obtain-the-access-token)
- you can store the access token and the refresh token on the backend
- the desktop app can then perform calls towards your backend, you can use the backend to pass the access token to the desktop app or use the backend as a wrapper of the FIC APIs to perform all the API calls directly from the BE, returning then the results to your desktop app. 

**Of course, you would need to implement proper authentication to your Server Side App.** In this case, the security of the app would be delegated to you.

The Access Token can be passed to the desktop app, the important thing is that you pass the correct access token to the user and that you expose the backend as HTTPS. If an attacker obtains an access token, only that user will be impacted and only until it expires, while if an attacker obtains the client_id and client_secret he will be able to require tokens impersonating your app. I would instead keep the Refresh token server-side because the refresh token endpoint still needs the client secret.

**The company ID is not a problem**. It is not sensible info (do not confuse it with the Client ID). You can keep or send it to the desktop app without a problem. Once you have the access token, you can use the [List User Companies](https://developers.fattureincloud.it/api-reference/#get-/user/companies) method to retrieve it, as explained [here](https://developers.fattureincloud.it/docs/basics/company-scoped-methods/).

Let us know your thoughts about it. 

Thanks,
Mauro

**Comment by ratataplan** [comment]
Yes, a cracker can obtains (very very difficult) the client_id and client_secret and he will be able to require tokens impersonating my app, but he must be a cracker **inside my organitation** because client_id and client_secret are stored in crypted mode in the DB of the organization.

Drawbacks of manual authentication:
- The User Experience is worse than the other authorization methods.
- The Error Ratio is higher because the user needs to select the scopes needed by the application; in contrast, in the other flows is the application that's indicating the scopes required to work properly.

Currently, the OAuth 2.0 Device Code Flow is not generally available; this means that you'll not be able to enable it from the Fatture in Cloud page

Thanks,
Carlo Mainardi

**Reply by valmoz** [reply]
Hi @ratataplan

1. Even if it is encrypted on your DB, your code needs to decrypt it to use it with our APIs, so your desktop app must have the decrypted version at least at that moment. So it is possible to obtain it, even if it is difficult
2. It is true that is not GA, but at the same time, I'm part of the team who can activate it... and we have other customers already using it in prod.
